pkg/planner/core/tiflash_predicate_push_down.go

```diff
@@ -0,0 +1,386 @@
+			return true
+		case ast.LogicAnd, ast.LogicOr:
+			return isPredicateSimpleCompare(sf.GetArgs()[0]) && isPredicateSimpleCompare(sf.GetArgs()[1])
+		case ast.UnaryNot:
+			return isPredicateSimpleCompare(sf.GetArgs()[0])
+		}
+	}
+	return false
+}
+
+// handleTiFlashPredicatePushDown is used to handle the TiFlash predicate push down.
+// 1. Whether to use the inverted index.
+// 2. Whether to push down the conditions to the table scan.
+func handleTiFlashPredicatePushDown(pctx base.PlanContext, ts *PhysicalTableScan, indexHints []*ast.IndexHint) {
+	// When the table is small, there is no need to push down the conditions.
+	if ts.tblColHists.RealtimeCount <= tiflashDataPackSize || ts.KeepOrder || len(ts.filterCondition) == 0 {
+		return
+	}
+
+	// Currently, TiFlash does not support vector index with predicates.
+	// So for now, one DataSource only contains one TiFlash !keepOrder path,
+	// which means that the following code will only be executed once.
+	// TODO: If there are multiple TiFlash !keepOrder paths in one DataSource in the future,
+	// we need to move the following code to another place.
+	// Since caculating selectivity is expensive, we need to avoid duplicate execution.
+	for _, index := range ts.UsedColumnarIndexes {
+		if index.IndexInfo.VectorInfo != nil {
+			panic("TiFlash does not support vector index with pedicates")
+		}
+	}
+
+	// Consider use index hints
+	indexMap := make(map[string]int, len(ts.Table.Indices))
+	for _, hint := range indexHints {
+		if hint.HintScope != ast.HintForScan {
+			continue
+		}
+
+		switch hint.HintType {
+		case ast.HintUse:
+			for _, name := range hint.IndexNames {
+				if indexMap[name.L] != math.MaxInt {
+					indexMap[name.L]++
+				}
+			}
+		case ast.HintIgnore:
+			for _, name := range hint.IndexNames {
+				if indexMap[name.L] != math.MaxInt {
+					indexMap[name.L]--
+				}
+			}
+		case ast.HintForce:
+			for _, name := range hint.IndexNames {
+				indexMap[name.L] = math.MaxInt
+			}
+		default:
+			continue
+		}
+	}
+
+	indexedColumnNameToIndexInfoMap := make(map[string]*model.IndexInfo, len(ts.Table.Indices))
+	for _, index := range ts.Table.Indices {
+		if index.State == model.StatePublic && index.InvertedInfo != nil {
+			if len(indexHints) > 0 && indexMap[index.Name.L] <= 0 {
+				continue
+			}
+			// inverted index only support one column
+			indexedColumnNameToIndexInfoMap[index.Columns[0].Name.L] = index
+		}
+	}
+
+	selectedColumns := make(map[string]bool, len(indexedColumnNameToIndexInfoMap))
+	selectedConditions := make([]expression.Expression, 0, len(ts.filterCondition))
+	for _, cond := range ts.filterCondition {
+		// 1. The predicate only contains >, >=, =, !=, <=, <, in.
+		if !isPredicateSimpleCompare(cond) {
+			continue
+		}
+		// 2. The columns of predicate all have inverted index.
+		allHaveInvertedIndex := true
+		columns := expression.ExtractColumns(cond)
+		columnNames := make([]string, 0, len(columns))
```

Comment:
Can we move this definition of columnNames outside this for-loop? We can reuse these slices to avoid allocating memory. remember to clear the slices at the end of for-loop